core/src/main/scala/org/apache/spark/rdd/PairRDDFunctions.scala

```diff
@@ -1196,28 +1198,40 @@ class PairRDDFunctions[K, V](self: RDD[(K, V)])
         writer.close()
       }
       writer.commit()
-      bytesWrittenCallback.foreach { fn => outputMetrics.setBytesWritten(fn()) }
-      outputMetrics.setRecordsWritten(recordsWritten)
+      outputMetrics.foreach { om =>
+        bytesWrittenCallback.foreach { fn => om.setBytesWritten(fn()) }
+        om.setRecordsWritten(recordsWritten)
+      }
     }
 
     self.context.runJob(self, writeToFile)
     writer.commitJob()
   }
 
-  private def initHadoopOutputMetrics(context: TaskContext): (OutputMetrics, Option[() => Long]) = {
+  // TODO: these don't seem like the right abstractions.
+  // We should abstract the duplicate code in a less awkward way.
+
+  private def initHadoopOutputMetrics(
+      context: TaskContext): (Option[OutputMetrics], Option[() => Long]) = {
```

Comment:
Question RE: return type: as far as I can tell, either both return fields will be defined or neither will. To make this more obvious, what about changing the return type to `Option[(OutputMetrics, () => Long)]` instead?